src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -260,16 +275,64 @@ int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t
                     gavl_left_rotation(cur_node, rnode);
             }
 
-            if (!STACK_EMPTY(node_stack)) {
-                STACK_POP(node_stack, cur_node);
-                goto stack_recovery;
+            if (!TREE_STACK_EMPTY(tree_ptr)) {
+                TREE_STACK_POP(tree_ptr, cur_node);
+                continue;
             } else {
                 break;
             }
         } while (1);
 
-        while (gavl_tree_iptr->root->parent != NULL)
-            gavl_tree_iptr->root = gavl_tree_iptr->root->parent;
+        while (tree_ptr->root && tree_ptr->root->parent != NULL)
+            tree_ptr->root = tree_ptr->root->parent;
+    }
+
+    return;
+}
+
+int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len,
+                         const void *val)
+{
+    int mpl_err = MPL_SUCCESS;
+    gavl_tree_node_s *node_ptr;
+    gavl_tree_s *tree_ptr = (gavl_tree_s *) gavl_tree;
+
+    node_ptr = (gavl_tree_node_s *) MPL_malloc(sizeof(gavl_tree_node_s), MPL_MEM_OTHER);
+    if (node_ptr == NULL) {
+        mpl_err = MPL_ERR_SHM_NOMEM;
+        goto fn_fail;
+    }
+
+    GAVL_TREE_NODE_INIT(node_ptr, addr, len, val);
+
+    if (tree_ptr->root == NULL) {
+        tree_ptr->root = node_ptr;
+    } else {
+        gavl_tree_node_s *pnode;
+        int direction;
+
+        /* search the node which will become the parent of new node */
+        gavl_tree_search_node(tree_ptr, (uintptr_t) node_ptr->addr, node_ptr->len,
+                              SUBSET_SEARCH, &pnode);
+
+        /* find which side the new node should be inserted */
+        direction = gavl_subset_cmp_func((uintptr_t) node_ptr->addr, node_ptr->len, pnode);
+        if (direction == BUFFER_MATCH) {
+            /* new node is duplicate, we need to delete new node and exit */
+            tree_ptr->gavl_free_fn((void *) node_ptr->val);
+            MPL_free(node_ptr);
+            goto fn_exit;
+        }
+
```

Comment:
This does not include the new design for replacing existing nodes with a superset new node, right?

Reply:
> Is access to node_stack thread safe? What happens if two threads call gavl concurrently?

I believe the entire gavl tree needs thread protection. Currently the entire shm path is protected by VCI(0) lock, so there is no need to worry about internal thread safety. @KaimingOuyang It is good put a comment at the top to note the shared global -- there's a global `struct gavl_tree` arrays somewhere, right? -- That'll make it easy when we consider multiple vci or more granular locks.

Reply:
no, this is just refactoring